File: src/sdk/typescript/src/invoke.ts
// 本ファイルは k1s0 TypeScript SDK の ServiceInvoke 動詞統一 facade。
// InvokeStream は本リリース時点 では raw 経由（client.transport から直接 PromiseClient を組む）。
import { createPromiseClient } from "@connectrpc/connect";
import type { K1s0Client } from "./client.js";
import { InvokeService } from "./proto/k1s0/tier1/serviceinvoke/v1/serviceinvoke_service_connect.js";

/** InvokeFacade は InvokeService の動詞統一 facade。 */
export class InvokeFacade {
  private readonly client: K1s0Client;

  constructor(client: K1s0Client) {
    this.client = client;
  }

  /** call は任意サービスの任意メソッドを呼び出す（unary）。 */
  async call(
    appId: string,
    method: string,
    data: Uint8Array,
    contentType: string,
    timeoutMs = 5000,
  ): Promise<{ data: Uint8Array; contentType: string; status: number }> {
    const raw = createPromiseClient(InvokeService, this.client.transport);
    const resp = await raw.invoke({
      appId,
      method,
      data,
      contentType,
      context: this.client.tenantContext(),
      timeoutMs,
    });
    return { data: resp.data, contentType: resp.contentType, status: resp.status };
  }
}

File: src/sdk/rust/crates/k1s0-sdk/src/invoke.rs
// 本ファイルは k1s0-sdk の ServiceInvoke 動詞統一 facade（unary + server streaming）。
use crate::client::Client;
use crate::proto::k1s0::tier1::serviceinvoke::v1::{
    InvokeChunk, InvokeRequest, invoke_service_client::InvokeServiceClient,
};
use tonic::{Status, Streaming, transport::Channel};

/// InvokeFacade は InvokeService の動詞統一 facade。
pub struct InvokeFacade {
    client: Client,
    raw: InvokeServiceClient<Channel>,
}

impl InvokeFacade {
    pub(crate) fn new(client: Client) -> Self {
        let raw = InvokeServiceClient::new(client.channel());
        Self { client, raw }
    }

    /// call は任意サービスの任意メソッドを呼び出す（unary）。返り値は (data, content_type, status)。
    pub async fn call(
        &mut self,
        app_id: &str,
        method: &str,
        data: Vec<u8>,
        content_type: &str,
        timeout_ms: i32,
    ) -> Result<(Vec<u8>, String, i32), Status> {
        let resp = self
            .raw
            .invoke(InvokeRequest {
                app_id: app_id.to_string(),
                method: method.to_string(),
                data,
                content_type: content_type.to_string(),
                context: Some(self.client.tenant_context()),
                timeout_ms,
            })
            .await?
            .into_inner();
        Ok((resp.data, resp.content_type, resp.status))
    }

    /// stream はサーバストリーミング呼出。`tonic::Streaming<InvokeChunk>` を返し、
    /// 利用者は `while let Some(chunk) = stream.message().await? { ... }` で消費する。
    pub async fn stream(
        &mut self,
        app_id: &str,
        method: &str,
        data: Vec<u8>,
        content_type: &str,
        timeout_ms: i32,
    ) -> Result<Streaming<InvokeChunk>, Status> {
        let resp = self
            .raw
            .invoke_stream(InvokeRequest {
                app_id: app_id.to_string(),
                method: method.to_string(),
                data,
                content_type: content_type.to_string(),
                context: Some(self.client.tenant_context()),
                timeout_ms,
            })
            .await?;
        Ok(resp.into_inner())
    }
}
